modules/core/shared/src/main/scala/io/circe/Error.scala

```diff
@@ -96,6 +123,13 @@ object DecodingFailure {
     val path = CursorOp.opsToPath(failure.history)
     s"DecodingFailure at ${path}: ${failure.message}"
   }
+
+  sealed trait Reason
+  case class MissingField(targetType: String) extends Reason
+  case class UnsupportedType(targetType: String) extends Reason
```

Comment:
No, IIRC the only instance of `UnsupportedType` is in CNil decoder.
I'd be more than happy to get rid of it, but I find `WrongTypeExpectation` not to be the right place since I don't know what the json type for a CNil should be (I don't even know why we need a decoder for CNil :))

Reply:
I feel like we don't if the derivation implicits work differently. @travisbrown might know exactly why, but if we provide a lower priority implicit for `H :+: CNil` (and not CNil by itself) it should still wrk right?

Either way, in reality CNil decoder's body will never be called so it doesn't matter what we put in it